messenger/decoder.go

```diff
@@ -0,0 +1,465 @@
+
+	// check "Connection" for "Keep-Alive"
+	d.kalive = d.req.Header.Get("Connection") == "Keep-Alive"
+
+	log.V(2).Infof(d.idtag+"update-for-request: chunked %v keep-alive %v", d.chunked, d.kalive)
+}
+
+func (d *httpDecoder) readBodyContent() httpState {
+	if d.chunked {
+		d.buf = &bytes.Buffer{}
+		return readChunkHeaderState
+	} else {
+		d.lrc = limit(d.rw.Reader, d.req.ContentLength)
+		d.buf = &bytes.Buffer{}
+		return readBodyState
+	}
+}
+
+const http202response = "HTTP/1.1 202 OK\r\nContent-Length: 0\r\n\r\n"
+
+func (d *httpDecoder) generateRequest() httpState {
+	log.V(2).Infof(d.idtag + "generate-request")
+	// send a Request to msg
+	b := d.buf.Bytes()
+	r := &http.Request{
+		Method:        d.req.Method,
+		URL:           d.req.URL,
+		Proto:         d.req.Proto,
+		ProtoMajor:    d.req.ProtoMajor,
+		ProtoMinor:    d.req.ProtoMinor,
+		Header:        d.req.Header,
+		Close:         !d.kalive,
+		Host:          d.req.Host,
+		RequestURI:    d.req.RequestURI,
+		Body:          &body{bytes.NewBuffer(b)},
+		ContentLength: int64(len(b)),
+	}
+
+	select {
+	case d.msg <- r:
+	case <-d.shouldQuit:
+		// best effort, we may have just tied
+		select {
+		case d.msg <- r:
+		default:
+		}
+		return terminateState
+	}
+
+	// send a 202 response to mesos so that it knows to continue
+	// TODO(jdef) set a write deadline here
+	_, err := d.rw.WriteString(http202response)
+	if err != nil {
+		if log.V(2) {
+			log.Errorln(d.idtag+"failed to write response code?!", err.Error())
+		}
+	} else {
+		err = d.rw.Flush()
+		if err != nil {
+			if log.V(2) {
+				log.Errorln(d.idtag+"failed to flush writer?!", err.Error())
+			}
+		}
+	}
+
+	// normally we'd terminate on the above error, but I'm testing something with libprocess: perhaps
+	// it leaves the connection half-open so that we can continue to receive events?
+
+	if d.kalive {
+		d.req = &http.Request{
+			ContentLength: -1,
+			Header:        make(http.Header),
+		}
+		return awaitRequestState
+	} else {
+		// TODO(jdef) we probably want to send a HTTP 202 or something?!
+		return terminateState
+	}
+}
+
+type httpState func(d *httpDecoder) httpState
+
+// terminateState shuts down the state machine
+func terminateState(d *httpDecoder) httpState {
+	close(d.msg)
+	close(d.errCh)
+	if d.con != nil {
+		d.con.Close()
+	}
+	return nil
+}
+
+type limitReadCloser struct {
+	*io.LimitedReader
+	r *bufio.Reader
+}
+
+func limit(r *bufio.Reader, limit int64) *limitReadCloser {
+	return &limitReadCloser{
+		LimitedReader: &io.LimitedReader{
+			R: r,
+			N: limit,
+		},
+		r: r,
+	}
+}
+
+func (l *limitReadCloser) Close() error {
+	// discard remaining
+	_, err := l.r.Discard(int(l.LimitedReader.N)) // TODO(jdef) possible overflow
+	return err
+}
+
+// bootstrapState expects to be called when the standard net/http lib has already
+// read the initial request query line + headers from a connection. the request
+// is ready to be hijacked at this point.
+func bootstrapState(d *httpDecoder) httpState {
```

Comment:
refactor this so that it's invoked DecodeHTTP and not actually a state. Then get rid of httpDecoder.res